File: oxidize-python/oxidize_python/core/model/gguf_decoder_load.py
"""GGUF weight loading for Llama decoder mirroring gguf_decoder_load.go."""

from __future__ import annotations

from dataclasses import dataclass

from oxidize_python.core.ggufcore.gguf import MappedFile, mapped_tensor_infos
from oxidize_python.core.model.dflash_weights import (
    F32Weight,
    f32_weight_from_dims,
    ones,
    quantized_gemv_supported,
    transpose_f32,
)
from oxidize_python.core.model.llama_decoder import (
    DecoderAttentionLayer,
    DecoderLayer,
    LlamaDecoderConfig,
    LlamaDecoderStack,
)
from oxidize_python.core.quantization import types as quant
from oxidize_python.core.quantization.dequant_k import dequantize
from oxidize_python.internal.gguf.types import TensorInfo


@dataclass
class _GgufWeightLoader:
    infos: list[TensorInfo]
    bytes_all: bytes
    config: LlamaDecoderConfig

    def _find(self, name: str) -> TensorInfo | None:
        for info in self.infos:
            if info.name == name:
                return info
        return None

    def load_f32_with_dims(self, name: str) -> tuple[list[float], list[int], bool]:
        info = self._find(name)
        if info is None:
            return [], [], False
        qtype = quant.from_ggml_type(info.ggml_type)
        count = 1
        for d in info.dimensions:
            count *= int(d)
        qsize = quant.quantized_size(qtype, count)
        off = info.absolute_offset
        end = off + qsize
        if end > len(self.bytes_all):
            raise ValueError(f"tensor {name} out of bounds")
        out = [0.0] * count
        dequantize(qtype, self.bytes_all[off:end], out)
        return out, list(info.dimensions), True

    def load_proj(self, name: str) -> F32Weight:
        info = self._find(name)
        if info is None:
            return F32Weight([], 0, 0)
        if len(info.dimensions) != 2:
            data, dims, ok = self.load_f32_with_dims(name)
            if not ok:
                return F32Weight([], 0, 0)
            return f32_weight_from_dims(data, dims)
        qtype = quant.from_ggml_type(info.ggml_type)
        in_dim = int(info.dimensions[0])
        out_dim = int(info.dimensions[1])
        if quantized_gemv_supported(qtype, in_dim):
            count = out_dim * in_dim
            qsize = quant.quantized_size(qtype, count)
            off = info.absolute_offset
            end = off + qsize
            if end > len(self.bytes_all):
                raise ValueError(f"tensor {name} out of bounds")
            raw = bytes(self.bytes_all[off:end])
            return F32Weight.from_quantized(raw, qtype, out_dim, in_dim)
        data, _, ok = self.load_f32_with_dims(name)
        if not ok:
            return F32Weight([], 0, 0)
        return F32Weight.from_slice(transpose_f32(data, in_dim, out_dim), out_dim, in_dim)

    def load_proj_any(self, *names: str) -> F32Weight:
        for name in names:
            w = self.load_proj(name)
            if w.is_loaded():
                return w
        return F32Weight([], 0, 0)

    def load_f32_any(self, *names: str) -> list[float] | None:
        for name in names:
            data, _, ok = self.load_f32_with_dims(name)[:3]
            if ok:
                return data
        return None


def load_llama_decoder_stack_from_gguf(
    mapped: MappedFile,
    config: LlamaDecoderConfig,
) -> tuple[LlamaDecoderStack, Exception | None]:
    try:
        stack = LlamaDecoderStack.new(config)
        loader = _GgufWeightLoader(
            mapped_tensor_infos(mapped.parsed),
            mapped.bytes,
            config,
        )
        norm = loader.load_f32_any("output_norm.weight", "norm.weight", "model.norm.weight")
        if norm:
            stack.norm = norm
        stack.output = loader.load_proj_any("lm_head.weight", "output.weight")
        stack.tok_embeddings = loader.load_proj_any(
            "model.embed_tokens.weight",
            "tok_embeddings.weight",
            "token_embd.weight",
        )
        if stack.tok_embeddings.is_loaded() and config.vocab_size == 0:
            stack.config.vocab_size = stack.tok_embeddings._output_dim()
        if not stack.output.is_loaded() and stack.tok_embeddings.is_loaded():
            stack.output = stack.tok_embeddings

        for layer_idx in range(config.layer_count):
            prefix = f"blk.{layer_idx}"
            in_ln = ones(config.hidden_size)
            data, _, ok = loader.load_f32_with_dims(f"{prefix}.attn_norm.weight")
            if ok:
                in_ln = data
            post_ln = ones(config.hidden_size)
            data, _, ok = loader.load_f32_with_dims(f"{prefix}.post_attention_norm.weight")
            if ok:
                post_ln = data
            else:
                data, _, ok = loader.load_f32_with_dims(f"{prefix}.ffn_norm.weight")
                if ok:
                    post_ln = data
            q_norm = ones(config.kv_head_dim() or config.head_dim())
            data, _, ok = loader.load_f32_with_dims(f"{prefix}.attn_q_norm.weight")
            if ok:
                q_norm = data
            k_norm = ones(config.kv_head_dim() or config.head_dim())
            data, _, ok = loader.load_f32_with_dims(f"{prefix}.attn_k_norm.weight")
            if ok:
                k_norm = data
            stack.layers.append(
                DecoderLayer(
                    input_layernorm=in_ln,
                    post_attention_layernorm=post_ln,
                    mlp_gate=loader.load_proj(f"{prefix}.ffn_gate.weight"),
                    mlp_up=loader.load_proj(f"{prefix}.ffn_up.weight"),
                    mlp_down=loader.load_proj(f"{prefix}.ffn_down.weight"),
                    attention=DecoderAttentionLayer(
                        q_proj=loader.load_proj(f"{prefix}.attn_q.weight"),
                        k_proj=loader.load_proj(f"{prefix}.attn_k.weight"),
                        v_proj=loader.load_proj(f"{prefix}.attn_v.weight"),
                        o_proj=loader.load_proj(f"{prefix}.attn_output.weight"),
                        q_norm_weight=q_norm,
                        k_norm_weight=k_norm,
                    ),
                )
            )
        return stack, None
    except Exception as err:
        return LlamaDecoderStack.new(config), err

File: oxidize-golang/core/model/gguf_decoder_load.go
package model

import (
	"fmt"

	"github.com/Zapdev-labs/oxidize/golang/core/ggufcore"
	"github.com/Zapdev-labs/oxidize/golang/core/quantization"
)

type ggufWeightLoader struct {
	infos    []ggufcore.TensorInfo
	bytesAll []byte
	config   LlamaDecoderConfig
}

func newGGUFWeightLoader(mapped *ggufcore.MappedFile, config LlamaDecoderConfig) *ggufWeightLoader {
	return &ggufWeightLoader{
		infos:    ggufcore.MappedTensorInfos(mapped.Parsed),
		bytesAll: mapped.Bytes,
		config:   config,
	}
}

func (l *ggufWeightLoader) loadF32WithDims(name string) ([]float32, []uint64, bool, error) {
	var info *ggufcore.TensorInfo
	for i := range l.infos {
		if l.infos[i].Name == name {
			info = &l.infos[i]
			break
		}
	}
	if info == nil {
		return nil, nil, false, nil
	}
	qtype := quantization.FromGGMLType(info.GGMLType)
	count := 1
	for _, d := range info.Dimensions {
		count *= int(d)
	}
	qsize, err := quantization.QuantizedSize(qtype, count)
	if err != nil {
		return nil, nil, false, err
	}
	off := int(info.AbsoluteOffset)
	end := off + qsize
	if end > len(l.bytesAll) {
		return nil, nil, false, fmt.Errorf("tensor %s out of bounds", name)
	}
	out := make([]float32, count)
	if err := quantization.DequantizeScalar(qtype, l.bytesAll[off:end], out); err != nil {
		return nil, nil, false, fmt.Errorf("dequantize %s: %w", name, err)
	}
	return out, info.Dimensions, true, nil
}

func (l *ggufWeightLoader) loadProj(name string) (F32Weight, error) {
	var info *ggufcore.TensorInfo
	for i := range l.infos {
		if l.infos[i].Name == name {
			info = &l.infos[i]
			break
		}
	}
	if info == nil {
		return F32Weight{}, nil
	}
	if len(info.Dimensions) != 2 {
		data, dims, ok, err := l.loadF32WithDims(name)
		if err != nil || !ok {
			return F32Weight{}, err
		}
		return f32WeightFromDims(data, dims), nil
	}
	qtype := quantization.FromGGMLType(info.GGMLType)
	inDim := int(info.Dimensions[0])
	outDim := int(info.Dimensions[1])
	if quantizedGemvSupported(qtype, inDim) {
		count := outDim * inDim
		qsize, err := quantization.QuantizedSize(qtype, count)
		if err != nil {
			return F32Weight{}, err
		}
		off := int(info.AbsoluteOffset)
		end := off + qsize
		if end > len(l.bytesAll) {
			return F32Weight{}, fmt.Errorf("tensor %s out of bounds", name)
		}
		return NewF32WeightFromQuantized(append([]byte(nil), l.bytesAll[off:end]...), qtype, outDim, inDim), nil
	}
	data, _, ok, err := l.loadF32WithDims(name)
	if err != nil || !ok {
		return F32Weight{}, err
	}
	return NewF32WeightFromSlice(transposeF32(data, inDim, outDim), outDim, inDim), nil
}

func (l *ggufWeightLoader) loadProjAny(names ...string) (F32Weight, error) {
	for _, name := range names {
		w, err := l.loadProj(name)
		if err != nil {
			return F32Weight{}, err
		}
		if w.IsLoaded() {
			return w, nil
		}
	}
	return F32Weight{}, nil
}

func (l *ggufWeightLoader) loadF32Any(names ...string) ([]float32, error) {
	for _, name := range names {
		data, _, ok, err := l.loadF32WithDims(name)
		if err != nil {
			return nil, err
		}
		if ok {
			return data, nil
		}
	}
	return nil, nil
}

// LoadLlamaDecoderStackFromGGUF loads tok embeddings, output head, norms, and blk.* layers.
func LoadLlamaDecoderStackFromGGUF(mapped *ggufcore.MappedFile, config LlamaDecoderConfig) (*LlamaDecoderStack, error) {
	stack := NewLlamaDecoderStack(config)
	loader := newGGUFWeightLoader(mapped, config)

	if norm, err := loader.loadF32Any("output_norm.weight", "norm.weight", "model.norm.weight"); err != nil {
		return nil, err
	} else if len(norm) > 0 {
		stack.Norm = norm
	}
	output, err := loader.loadProjAny("lm_head.weight", "output.weight")
	if err != nil {
		return nil, err
	}
	if output.IsLoaded() {
		stack.Output = output
	}
	tokEmb, err := loader.loadProjAny("model.embed_tokens.weight", "tok_embeddings.weight", "token_embd.weight")
	if err != nil {
		return nil, err
	}
	if tokEmb.IsLoaded() {
		stack.TokEmbeddings = tokEmb
		if config.VocabSize == 0 {
			stack.Config.VocabSize = tokEmb.outputDim()
		}
	}
	if !stack.Output.IsLoaded() && stack.TokEmbeddings.IsLoaded() {
		stack.Output = stack.TokEmbeddings
	}

	for layerIdx := 0; layerIdx < config.LayerCount; layerIdx++ {
		prefix := fmt.Sprintf("blk.%d", layerIdx)
		inLN := ones(config.HiddenSize)
		if data, _, ok, _ := loader.loadF32WithDims(prefix + ".attn_norm.weight"); ok {
			inLN = data
		}
		postLN := ones(config.HiddenSize)
		if data, _, ok, _ := loader.loadF32WithDims(prefix + ".post_attention_norm.weight"); ok {
			postLN = data
		} else if data, _, ok, _ := loader.loadF32WithDims(prefix + ".ffn_norm.weight"); ok {
			postLN = data
		}
		// Gemma sandwich norm: post_attention_norm is the true post-attention
		// norm, ffn_norm is the pre-MLP norm, and post_ffw_norm is the post-FFN
		// norm. Load them separately so the forward pass can apply all four.
		var preFFN, postFFN []float32
		if config.SandwichNorm {
			if data, _, ok, _ := loader.loadF32WithDims(prefix + ".post_attention_norm.weight"); ok {
				postLN = data
			}
			if data, _, ok, _ := loader.loadF32WithDims(prefix + ".ffn_norm.weight"); ok {
				preFFN = data
			} else {
				preFFN = ones(config.HiddenSize)
			}
			if data, _ := loader.loadF32Any(prefix+".post_ffw_norm.weight", prefix+".post_ffn_norm.weight"); len(data) > 0 {
				postFFN = data
			} else {
				postFFN = ones(config.HiddenSize)
			}
		}
		qProj, err := loader.loadProj(prefix + ".attn_q.weight")
		if err != nil {
			return nil, err
		}
		kProj, err := loader.loadProj(prefix + ".attn_k.weight")
		if err != nil {
			return nil, err
		}
		vProj, err := loader.loadProj(prefix + ".attn_v.weight")
		if err != nil {
			return nil, err
		}
		oProj, err := loader.loadProj(prefix + ".attn_output.weight")
		if err != nil {
			return nil, err
		}
		qNorm := ones(config.KVHeadDim())
		if config.KVHeadDim() == 0 {
			qNorm = ones(config.HeadDim())
		}
		if data, _, ok, _ := loader.loadF32WithDims(prefix + ".attn_q_norm.weight"); ok {
			qNorm = data
		}
		kNorm := ones(config.KVHeadDim())
		if config.KVHeadDim() == 0 {
			kNorm = ones(config.HeadDim())
		}
		if data, _, ok, _ := loader.loadF32WithDims(prefix + ".attn_k_norm.weight"); ok {
			kNorm = data
		}
		gate, err := loader.loadProj(prefix + ".ffn_gate.weight")
		if err != nil {
			return nil, err
		}
		up, err := loader.loadProj(prefix + ".ffn_up.weight")
		if err != nil {
			return nil, err
		}
		down, err := loader.loadProj(prefix + ".ffn_down.weight")
		if err != nil {
			return nil, err
		}
		moe := MoELayer{NumExperts: config.NumExperts}
		if config.NumExperts > 0 {
			gateInp, err := loader.loadProj(prefix + ".ffn_gate_inp.weight")
			if err != nil {
				return nil, err
			}
			gateExps, err := loader.loadProj(prefix + ".ffn_gate_exps.weight")
			if err != nil {
				return nil, err
			}
			upExps, err := loader.loadProj(prefix + ".ffn_up_exps.weight")
			if err != nil {
				return nil, err
			}
			downExps, err := loader.loadProj(prefix + ".ffn_down_exps.weight")
			if err != nil {
				return nil, err
			}
			moe = MoELayer{
				GateInp:    gateInp,
				GateExps:   gateExps,
				UpExps:     upExps,
				DownExps:   downExps,
				NumExperts: config.NumExperts,
			}
		}
		stack.Layers = append(stack.Layers, DecoderLayer{
			InputLayernorm:         inLN,
			PostAttentionLayernorm: postLN,
			PreFFNLayernorm:        preFFN,
			PostFFNLayernorm:       postFFN,
			MLPGate:                gate,
			MLPUp:                  up,
			MLPDown:                down,
			MoE:                    moe,
			Attention: DecoderAttentionLayer{
				QProj:       qProj,
				KProj:       kProj,
				VProj:       vProj,
				OProj:       oProj,
				QNormWeight: qNorm,
				KNormWeight: kNorm,
			},
		})
	}
	return stack, nil
}
